File: app/src/main/java/com/lacour/vincent/hypnosedetente/service/ForegroundService.java
package com.lacour.vincent.hypnosedetente.service;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.app.Service;
import android.content.Intent;
import android.os.Build;
import android.os.Bundle;
import android.os.IBinder;
import android.support.v4.app.NotificationCompat;

import com.lacour.vincent.hypnosedetente.R;
import com.lacour.vincent.hypnosedetente.screen.AudioPlayer;

public class ForegroundService extends Service {

    public static final String ACTION_START_FOREGROUND_SERVICE = "ACTION_START_FOREGROUND_SERVICE";
    public static final String ACTION_STOP_FOREGROUND_SERVICE = "ACTION_STOP_FOREGROUND_SERVICE";

    final String NOTIFICATION_CHANNEL_ID = "57070615ece54925b49fc1ce1cb965b3";
    final CharSequence NOTIFICATION_CHANNEL_NAME = "COM.LACOUR.VINCENT.HYPNOSEDETENTE";
    final String NOTIFICATION_CHANNEL_DESCRIPTION = "FOREGROUND_SERVICE";
    final int FOREGROUND_SERVICE_ID = 20100;


    @Override
    public IBinder onBind(Intent intent) {
        // TODO: Return the communication channel to the service.
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public void onCreate() {
        super.onCreate();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (intent != null) {
            String action = intent.getAction();
            Bundle bundle = intent.getExtras();
            String selectedSound = "";
            if (bundle != null) {
                selectedSound = bundle.getString("SELECTED_SOUND");
            }
            if (action != null) {
                switch (action) {
                    case ACTION_START_FOREGROUND_SERVICE:
                        this.startForegroundService(selectedSound);
                        break;
                    case ACTION_STOP_FOREGROUND_SERVICE:
                        this.stopForegroundService();
                        break;
                    default:
                        this.stopForegroundService();
                        break;
                }
            }
        }
        return super.onStartCommand(intent, flags, startId);
    }

    private void startForegroundService(String selectedSound) {
        Intent notificationIntent = new Intent(this, AudioPlayer.class);
        notificationIntent.addFlags(Intent.FLAG_ACTIVITY_SINGLE_TOP);
        PendingIntent pendingIntent = PendingIntent.getActivity(this, (int) System.currentTimeMillis(), notificationIntent, PendingIntent.FLAG_UPDATE_CURRENT);

        createNotificationChannel();
        NotificationCompat.Builder builder = new NotificationCompat.Builder(this, NOTIFICATION_CHANNEL_ID);
        builder.setContentTitle(getString(R.string.appName));
        builder.setContentText(getString(R.string.notificationContent, selectedSound));
        builder.setWhen(System.currentTimeMillis());
        builder.setSmallIcon(R.drawable.ic_music_note);
        builder.setPriority(Notification.PRIORITY_LOW);
        builder.setContentIntent(pendingIntent);
        Notification notification = builder.build();

        this.startForeground(FOREGROUND_SERVICE_ID, notification);
    }

    private void createNotificationChannel() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            NotificationChannel channel = new NotificationChannel(NOTIFICATION_CHANNEL_ID, NOTIFICATION_CHANNEL_NAME, NotificationManager.IMPORTANCE_DEFAULT);
            channel.setDescription(NOTIFICATION_CHANNEL_DESCRIPTION);
            NotificationManager notificationManager = getSystemService(NotificationManager.class);
            notificationManager.createNotificationChannel(channel);
        }
    }

    private void stopForegroundService() {
        this.stopForeground(true);
        stopSelf();
    }
}

File: app/src/main/java/com/lacour/vincent/hypnosedetente/service/ForegroundService.kt
package com.lacour.vincent.hypnosedetente.service

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.PendingIntent
import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.IBinder
import androidx.core.app.NotificationCompat

import com.lacour.vincent.hypnosedetente.screen.AudioPlayer
import com.lacour.vincent.hypnosedetente.R


class ForegroundService : Service() {

    companion object {
        const val ACTION_START_FOREGROUND_SERVICE: String = "ACTION_START_FOREGROUND_SERVICE"
        const val ACTION_STOP_FOREGROUND_SERVICE = "ACTION_STOP_FOREGROUND_SERVICE"

        private const val NOTIFICATION_CHANNEL_ID: String = "57070615ece54925b49fc1ce1cb965b3hgtff"
        private val NOTIFICATION_CHANNEL_NAME: CharSequence = "COM.LACOUR.VINCENT.HYPNOSEDETENTE"
        private const val NOTIFICATION_CHANNEL_DESCRIPTION: String = "FOREGROUND_SERVICE"
        private const val FOREGROUND_SERVICE_ID: Int = 20100
    }


    override fun onBind(intent: Intent): IBinder? {
        // TODO: Return the communication channel to the service.
        throw UnsupportedOperationException("Not yet implemented")
    }

    override fun onDestroy() {
        super.onDestroy()
        stopForeground(true)
    }

    override fun onTaskRemoved(rootIntent: Intent?) {
        super.onTaskRemoved(rootIntent)
        stopSelf()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (intent != null) {
            val action = intent.action
            val bundle = intent.extras
            val selectedSound: String = if (bundle != null) bundle.getString("SELECTED_SOUND")!! else ""
            if (action != null) {
                when (action) {
                    ACTION_START_FOREGROUND_SERVICE -> getNotification(selectedSound)
                    ACTION_STOP_FOREGROUND_SERVICE -> stopForegroundService()
                    else -> this.stopForegroundService()
                }
            }
        }
        return super.onStartCommand(intent, flags, startId)
    }

    private fun getNotification(seletedSound: String) {

        createChannel(this)

        val notifyIntent = Intent(this, AudioPlayer::class.java)
        notifyIntent.flags = Intent.FLAG_ACTIVITY_SINGLE_TOP
        val pendingIntent = PendingIntent.getActivity(
            this, 0, notifyIntent, PendingIntent.FLAG_UPDATE_CURRENT
        )

        val mNotification: Notification = NotificationCompat.Builder(this, NOTIFICATION_CHANNEL_ID)
            .setSmallIcon(R.drawable.ic_music_note)
            .setWhen(System.currentTimeMillis())
            .setAutoCancel(true)
            .setContentTitle(getString(R.string.app_name))
            .setContentText(getString(R.string.notification_description, seletedSound))
            .setContentIntent(pendingIntent)
            .build()

        startForeground(FOREGROUND_SERVICE_ID, mNotification)
    }


    private fun createChannel(context: Context) {

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {

            // Create the NotificationChannel, but only on API 26+ because
            // the NotificationChannel class is new and not in the support library
            val notificationManager = context.getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            val importance = NotificationManager.IMPORTANCE_LOW
            val notificationChannel =
                NotificationChannel(NOTIFICATION_CHANNEL_ID, NOTIFICATION_CHANNEL_NAME, importance)
            notificationChannel.enableVibration(true)
            notificationChannel.setShowBadge(true)
            notificationChannel.enableLights(true)
            notificationChannel.description = NOTIFICATION_CHANNEL_DESCRIPTION
            notificationChannel.lockscreenVisibility = Notification.VISIBILITY_PUBLIC
            notificationManager.createNotificationChannel(notificationChannel)
        }

    }

    private fun stopForegroundService() {
        stopForeground(true)
        stopSelf()
    }

}